pkg/asset/agent/image/agentpxefiles.go

```diff
@@ -0,0 +1,178 @@
+package image
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/diskfs/go-diskfs/filesystem"
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/assisted-image-service/pkg/isoeditor"
+	"github.com/openshift/installer/pkg/asset"
+)
+
+const (
+	initrdimg = "initrd"
+	rootfsimg = "rootfs"
+	vmlinuz   = "vmlinuz"
+	// pxeAssetsPath is the path where pxe files are created.
+	pxeAssetsPath = "pxe"
+)
+
+// AgentPXEFiles is an asset that generates the bootable image used to install clusters.
+type AgentPXEFiles struct {
+	imageReader isoeditor.ImageReader
+	cpuArch     string
+	isoPath     string
+}
+
+var _ asset.WritableAsset = (*AgentPXEFiles)(nil)
+
+// Dependencies returns the assets on which the AgentPXEFiles asset depends.
+func (a *AgentPXEFiles) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&Ignition{},
+		&BaseIso{},
+	}
+}
+
+// Generate generates the image files for PXE asset.
+func (a *AgentPXEFiles) Generate(dependencies asset.Parents) error {
+	ignition := &Ignition{}
+	dependencies.Get(ignition)
+
+	baseImage := &BaseIso{}
+	dependencies.Get(baseImage)
+
+	a.isoPath = baseImage.baseISOPath
+
+	tmpdir, err := os.MkdirTemp("", pxeAssetsPath)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpdir)
+
+	srcfilename := fmt.Sprintf("images/pxeboot/%s.img", initrdimg)
+	dstfilename := filepath.Join(tmpdir, fmt.Sprintf("%s.img", initrdimg))
+	err = a.extractPXEFileFromISO(a.isoPath, srcfilename, dstfilename)
+	if err != nil {
+		return err
+	}
+
+	ignitionByte, err := json.Marshal(ignition.Config)
+	if err != nil {
+		return err
+	}
+
+	ignitionContent := &isoeditor.IgnitionContent{Config: ignitionByte}
+
+	fname := filepath.Join(tmpdir, fmt.Sprintf("%s.img", initrdimg))
+	custom, err := isoeditor.NewInitRamFSStreamReader(fname, ignitionContent)
+	if err != nil {
+		return err
+	}
+
+	a.imageReader = custom
+	a.cpuArch = ignition.CPUArch
+
+	return nil
+}
+
+// PersistToFile writes the PXE assets in the assets folder named pxe.
+func (a *AgentPXEFiles) PersistToFile(directory string) error {
+	// If the imageReader is not set then it means that either one of the AgentPXEFiles
+	// dependencies or the asset itself failed for some reason
+	if a.imageReader == nil {
+		return errors.New("cannot generate PXE assets due to configuration errors")
+	}
+
+	defer a.imageReader.Close()
+	pxeAssetsFullPath := filepath.Join(directory, pxeAssetsPath)
+
+	os.RemoveAll(pxeAssetsFullPath)
+
+	err := os.Mkdir(pxeAssetsFullPath, 0750)
+	if err != nil {
+		return err
+	}
+
+	agentInitrdFile := filepath.Join(pxeAssetsFullPath, fmt.Sprintf("agent-%s.%s.img", initrdimg, a.cpuArch))
+	output, err := os.Create(agentInitrdFile)
```

Comment:
You could use the copy() function below that creates the file and copies data to it.